futures-util/src/async_await/select.rs

```diff
@@ -19,47 +26,104 @@
 /// let mut b = future::empty::<()>();
 ///
 /// let res = select! {
-///     future(a as a) => a + 1,
-///     future(b as _) => 0,
+///     a_res = a => a_res + 1,
+///     _ = b => 0,
 /// };
 /// assert_eq!(res, 5);
 /// # });
 /// ```
 ///
-/// In addition to `future(...)` matchers, `select!` accepts an `all complete`
-/// branch which can be used to match on the case where all the futures passed
-/// to select have already completed.
+/// ```
+/// #![feature(pin, async_await, await_macro, futures_api)]
+/// # futures::executor::block_on(async {
+/// use futures::future::{self, FutureExt};
+/// use futures::stream::{self, StreamExt};
+/// use futures::select;
+/// let mut st = stream::iter(vec![2]).fuse();
+/// let mut fut = future::empty::<()>();
+///
+/// select! {
+///     x = st.next() => assert_eq!(Some(2), x),
+///     _ = fut => panic!(),
+/// };
+/// # });
+/// ```
+///
+/// `select` also accepts a `complete` branch and a `default` branch.
+/// `complete` will run if all futures and streams have already been
+/// exhausted. `default` will run if no futures or streams are
+/// immediately ready.
```

Comment:
`default` does have lower priority than `complete`. WRT order, I considered it-- I don't feel super strongly about it either way.

Reply:
Me neither. But I think at the very least we should put `default` after `complete` in the documented examples.

Reply:
oh, that's funny-- I had been thinking of it in terms of "default runs before complete", not "complete has higher priority over default"

Reply:
Ah, interesting! So our intuitions differ in chronological order vs evaluation order.

In my view, `select!` evaluates similarly to `match` - it goes from top to bottom until it finds the first pattern that matches. If you think about it that way, and write a `default` case before `complete`, then it kinda looks like `complete` has no chance of ever running because `default` will get matched first. :)

To draw a parallel with `match`:

```rust
match 1 {
    _ => println!("default"),
    x if x > 0 => println!("complete"), // unreachable pattern :(
}
```

Reply:
Yup, makes sense! I'll adjust the order in the examples